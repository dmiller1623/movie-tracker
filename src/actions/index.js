export const populateMovies = (movies) => ({
  type: 'POPULATE_MOVIES',
  movies
});

export const loginUser = (user) => ({
  type: 'LOGIN_USER',
  user
});

export const signUpUser = (email, password) => ({
  type: 'SIGNUP_USER',
  email,
  password
});

export const signOutUser = () => {
  type: 'SIGN_OUT'

}
