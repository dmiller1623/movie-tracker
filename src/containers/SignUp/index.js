import React, { Component } from 'react';
import { connect } from 'react-redux';

import { signUpUser } from '../../actions';
import { addNewUser, getUser } from '../../utilities/apiCalls/apiCalls';
import PropTypes from 'prop-types';

export class SignUp extends Component {
  constructor() {
    super();
    this.state = {
      name: '',
      email: '',
      password: ''
    };
  }

  handleChange = (event) => {
    const { name, value } = event.target;
    this.setState({[name]: value});
  }

  submitSignUp = async (event) => {
    event.preventDefault();
    const { name, email, password } = this.state;
    if (name && email && password) {
      const cleanEmail = email.trim().toLowerCase();
      const capitalName = name.charAt(0).toUpperCase() + name.slice(1);
      await addNewUser(capitalName, cleanEmail, password);
      const newUser = await getUser(cleanEmail, password);
      this.props.signUpUser(newUser);
      this.props.history.push('/');
    } else {
      alert('Please enter a name, email, and password');
    }
  }

  render() {
    const { email, password, name } = this.state;

    return (
      <div>
        <form onSubmit={this.submitSignUp}>
          <input
            name='name'
            value={name}
            placeholder='name'
            autoComplete="on"
            onChange={this.handleChange}
          />
          <input
            type='email'
            name='email'
            value={email}
            placeholder='email'
            autoComplete="on"
            onChange={this.handleChange}
          />
          <input
            name='password'
            type='password'
            value={password}
            placeholder='password'
            autoComplete="off"
            onChange={this.handleChange}
          />
          <button>Sign Up</button>
        </form>
      </div>
    );
  }
}

export const mapDispatchToProps = (dispatch) => ({
  signUpUser: (user) => dispatch(signUpUser(user))
});

SignUp.propTypes = {
  signUpUser: PropTypes.func,
  history: PropTypes.array
};

export default connect(null, mapDispatchToProps)(SignUp);
