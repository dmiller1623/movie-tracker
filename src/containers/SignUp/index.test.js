import React from 'react'
import { shallow, mount } from 'enzyme'
import {  SignUp, mapDispatchToProps } from '../SignUp'
import { signUpUser } from '../../actions'


describe('SignUp', () => {
  let wrapper
  describe('SignUp component', () => {
    wrapper = mount(<SignUp />)

    it('should match the snapshot', () => {
      wrapper = shallow(<SignUp />)
      expect(wrapper).toMatchSnapshot()
    })
    it('should update state when user enters something in the name input', () => {
      const mockEvent = { target: { value: 'Dennis', name: 'name' } }
      const expectedState = {
        name: 'Dennis',
        email: '', 
        password: ''
      };
      wrapper.instance().handleChange(mockEvent)
      expect(wrapper.state()).toEqual(expectedState);
    })

    it('should update state when user enters something in the email input', () => {
      const mockEvent = { target: { value: 'dmill@aol.com', name: 'email' } }
      const expectedState = {
        name: 'Dennis',
        email: 'dmill@aol.com', 
        password: ''
      };
      wrapper.instance().handleChange(mockEvent)
      expect(wrapper.state()).toEqual(expectedState);
    })

    it('should update state when user enters something in the paswword input', () => {
      const mockEvent = { target: { value: 'password', name: 'password' } }
      const expectedState = {
        name: 'Dennis',
        email: 'dmill@aol.com', 
        password: 'password'
      };
      wrapper.instance().handleChange(mockEvent)
      expect(wrapper.state()).toEqual(expectedState);
    })

    it('should call submitSignUp on form submit', async () => {
      window.fetch = jest.fn().mockImplementation(() => 
      Promise.resolve({
        ok: true, 
        json: () => Promise.resolve({ name:'Dennis', email:'dm@aol.com', password:'password' })
      }));
      const mockEvent = { preventDefault: jest.fn() }
      const mockSignUp = jest.fn()
      const history = []
      wrapper = shallow(<SignUp signUpUser={mockSignUp} history={history}/>)
      await wrapper.instance().submitSignUp(mockEvent)
      expect(mockSignUp).toHaveBeenCalled()
    })
  })

  describe('matchdispatchtoprops', () => {
    it('should call dispatch when using a function MDTP', () => {
      const user = {
        email: 'dm@aol.com',
        id:1,
        name: 'Dennis',
        password: 'password'
      }

      const mockDispatch = jest.fn()
      const actionToDispatch = signUpUser(user)

      const mappedProps = mapDispatchToProps(mockDispatch)

      mappedProps.signUpUser(user)

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)

    })
  })
})