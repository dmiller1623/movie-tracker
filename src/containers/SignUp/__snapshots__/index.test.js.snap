// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp SignUp component should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SignUp />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <form
        onSubmit={[Function]}
>
        <input
                autoComplete="on"
                name="name"
                onChange={[Function]}
                placeholder="name"
                value=""
        />
        <input
                autoComplete="on"
                name="email"
                onChange={[Function]}
                placeholder="email"
                type="email"
                value=""
        />
        <input
                autoComplete="off"
                name="password"
                onChange={[Function]}
                placeholder="password"
                type="password"
                value=""
        />
        <button>
                Sign Up
        </button>
</form>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            autoComplete="on"
            name="name"
            onChange={[Function]}
            placeholder="name"
            value=""
/>,
          <input
            autoComplete="on"
            name="email"
            onChange={[Function]}
            placeholder="email"
            type="email"
            value=""
/>,
          <input
            autoComplete="off"
            name="password"
            onChange={[Function]}
            placeholder="password"
            type="password"
            value=""
/>,
          <button>
            Sign Up
</button>,
        ],
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "autoComplete": "on",
            "name": "name",
            "onChange": [Function],
            "placeholder": "name",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "autoComplete": "on",
            "name": "email",
            "onChange": [Function],
            "placeholder": "email",
            "type": "email",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "autoComplete": "off",
            "name": "password",
            "onChange": [Function],
            "placeholder": "password",
            "type": "password",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Sign Up",
          },
          "ref": null,
          "rendered": "Sign Up",
          "type": "button",
        },
      ],
      "type": "form",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <form
          onSubmit={[Function]}
>
          <input
                    autoComplete="on"
                    name="name"
                    onChange={[Function]}
                    placeholder="name"
                    value=""
          />
          <input
                    autoComplete="on"
                    name="email"
                    onChange={[Function]}
                    placeholder="email"
                    type="email"
                    value=""
          />
          <input
                    autoComplete="off"
                    name="password"
                    onChange={[Function]}
                    placeholder="password"
                    type="password"
                    value=""
          />
          <button>
                    Sign Up
          </button>
</form>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              autoComplete="on"
              name="name"
              onChange={[Function]}
              placeholder="name"
              value=""
/>,
            <input
              autoComplete="on"
              name="email"
              onChange={[Function]}
              placeholder="email"
              type="email"
              value=""
/>,
            <input
              autoComplete="off"
              name="password"
              onChange={[Function]}
              placeholder="password"
              type="password"
              value=""
/>,
            <button>
              Sign Up
</button>,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "autoComplete": "on",
              "name": "name",
              "onChange": [Function],
              "placeholder": "name",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "autoComplete": "on",
              "name": "email",
              "onChange": [Function],
              "placeholder": "email",
              "type": "email",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "autoComplete": "off",
              "name": "password",
              "onChange": [Function],
              "placeholder": "password",
              "type": "password",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Sign Up",
            },
            "ref": null,
            "rendered": "Sign Up",
            "type": "button",
          },
        ],
        "type": "form",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
